### PR TITLE
cody web: do not include ignored files in context

### PR DESCRIPTION
**Pull request by taras-yemets**
Excludes files from Cody scope if they match patterns from the Cody ignore file.

**Ignored file content** 👇🏻 
Screenshot 2024-01-29 at 15 55 53

| Ignored | Chat | Transcript |
| -- | -- | -- |
| ✅  |Screenshot 2024-01-29 at 15 54 51|Screenshot 2024-01-29 at 16 00 20|
| 🚫 |Screenshot 2024-01-29 at 15 55 25|Screenshot 2024-01-29 at 15 57 15|

> [!note]
> Known issue: Cody chat refers to the previously opened file in the current transcript if the current file is excluded from the scope (https://github.com/sourcegraph/cody/issues/2925). FYI when testing this PR. 

## Test plan
- Open repo which has `.cody/ignore` file
- Open a file matching the pattern from the ignore file
- View chat transcript (e.g., network tab) and ensure that code snippets from the ignored file are not included

### CODE REVIEW

File: client/web/src/cody/components/ScopeSelector/RepositoriesSelectorPopover.tsx [hunk 1]
```diff
@@ -214,6 +214,7 @@ export const RepositoriesSelectorPopover: React.FC<{
                                     <div className={classNames('d-flex flex-column', styles.contextItemsContainer)}>
                                         {inferredRepository && (
                                             <div className="d-flex flex-column py-1">
+                                                {/* TODO: ensure it's not rendered */}
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/cody/components/ScopeSelector/RepositoriesSelectorPopover.tsx]: 😅 Ensured - it's not rendered
https://github.com/sourcegraph/sourcegraph/pull/59907/commits/efa3a355570fe125cfb18cb74ca35257a4208559

**Comment by taras-yemets** [comment]
>Have we already hooked the same check at backend for files returned on fetching context using getCodyContext?

- https://github.com/sourcegraph/sourcegraph/pull/59836 

cc: @thenamankumar

**Comment by chwarwick** [comment]
> I see we are using `useCodyIgnore` only to check the currently opened file path. Have we already hooked the same check at backend for files returned on fetching context using `getCodyContext`?

@thenamankumar I'm working on updating the `getCodyContext` api currently.

**Comment by chwarwick** [comment]
> @chwarwick, is the current plan to put checking the ignore file behind the feature flag (https://github.com/sourcegraph/sourcegraph/pull/59836/files#r1465507213)?

Yes I was going to add an experimental setting for it since that aligns with vs code currently.

**Comment by taras-yemets** [comment]
> Yes I was going to add an experimental setting for it since that aligns with vs code currently.

@chwarwick, we should probably respect this setting on the client too. I'll hold on merging this PR until yours is merged.

File: client/web/src/cody/components/ChatUI/useCodyIgnore.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                    content
+                }
+            }
+        }
+    }
+`
+
+const CODY_IGNORE_PATH = '.cody/ignore'
+
+export const useCodyIgnore = (): { ignores: (path: string) => boolean } => {
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/cody/components/ChatUI/useCodyIgnore.ts]: https://github.com/sourcegraph/sourcegraph/pull/59907/commits/dbfbee1cff2cab7d069fb34e638b2babf923a636

**Review by thenamankumar — APPROVED** [review]

**Review by arafatkatze — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -25,6 +25,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Expiry to access tokens. Users can now select a maximum timespan for which a token is valid. Tokens will automatically lose access after this period. Default timeframes and an override to allow access tokens without expiration can be configured in the `auth.accessTokens` section of the site configuration. [#59565](https://github.com/sourcegraph/sourcegraph/pull/59565)
 - Gerrit code host connections now support an 'exclude' field that prevents repos in this list from being synced. [#59739](https://github.com/sourcegraph/sourcegraph/pull/59739)
 - Limit the number of active access tokens for a user. By default users are able to have 25 active access tokens. This limit can be configured using the `maxTokensPerUser` setting in the `auth.accessTokens` section of the site configuration. [#59731](https://github.com/sourcegraph/sourcegraph/pull/59731)
+- Add .cody/ignore support to remote context. To enable it, set `experimentalFeatures.codyContextIgnore: true` in the site configuration. [#59836](https://github.com/sourcegraph/sourcegraph/pull/59836), [#59907](https://github.com/sourcegraph/sourcegraph/pull/59907)
```
Comment (chwarwick) [inline CHANGELOG.md:28]:
```suggestion
- Add experimental support for .cody/ignore when retrieving remote context. To enable it, set `experimentalFeatures.codyContextIgnore: true` in the site configuration. [#59836](https://github.com/sourcegraph/sourcegraph/pull/59836), [#59907](https://github.com/sourcegraph/sourcegraph/pull/59907)
```

**Review by chwarwick — APPROVED** [review]

**Pull request merged**